blog/content/post/2025-01-21-ram-data-decay-research-part2.md

```diff
@@ -0,0 +1,489 @@
+
+{{< figure src="/img/ram_remanence_plots/out_0.2.png"
+caption="Bitmap with BER = 0.2" >}}
+
+{{< figure src="/img/ram_remanence_plots/out_0.1.png"
+caption="Bitmap with BER = 0.1" >}}
+
+<!-- markdownlint-disable MD013 -->
+{{< figure src="/img/ram_remanence_plots/out_0.0.png"
+caption="Bitmap with BER = 0.0 - original image. By: [Finlay McWalter, Public Domain](https://commons.wikimedia.org/w/index.php?curid=68100)" >}}
+<!-- markdownlint-restore -->
+
+> Note: BER was simulated for pixel data only, and not metadata like image
+> dimensions, pixel format or any possible checksum. Doing so would most likely
+> cause the decoders to detect the file as damaged and refuse displaying it
+> further.
+
+With BER = 50%, all we get is white noise. With 40% error ratio, some edges
+become visible, but depending on the complexity of the image, this may or may
+not be enough to recognize the objects. Teapot on images with lower BER values
+is clearly recognizable, despite the noise.
+
+This also is a good moment to show why 50% is the worst case. With higher error
+rate, the resulting bits are actually more correlated with the source material,
+but inverted. For example, image with BER = 95% holds the same amount of data as
+5% would, but all the colors are inverted:
+
+{{< figure src="/img/ram_remanence_plots/out_0.95.png"
+caption="Bitmap with BER = 0.95" >}}
+
+The same was shown for text with BER = 100%. Even though not a single output
+character was printable, each byte can be inverted to restore original message
+without any losses.
+
+## Summary
+
+There is no clear limit as to the point after which memory content has "decayed
+enough". Some data is easier to recover from partial structures than others. In
+some cases, like bitmaps shown above, losing some bits is acceptable, and while
+it degrades the quality of the image, the message stored within can still be
+recovered. This doesn't apply to all of the data in memory, in some cases each
+bit is significant. If the structure has a checksum, it may be used to help with
+guessing the proper values, but as such checksum would also be susceptible to
+decay, it isn't always a reliable solution.
+
+For files that are somehow encoded (e.g. compressed or encrypted), a single
+error in one data run may potentially break the ability to parse further
+structures. However, encoded or encrypted data in RAM is rare - it is used to
+store files on disk, but reading them to memory often converts it to raw,
+uncompressed and decrypted format.
+
+As was shown, the time for which data is preserved in RAM after the platform is
+powered down depends on many factors, but in general there should be not enough
+useful data to recover after about 3 seconds on most of the tested modules. A
+notable exception is 4 GB Kingston Fury KF432C16BB/4, where similar level of
+memory content loss was measured after about a minute. Interestingly, such
+results weren't reproduced for other DDR4 modules, even from the same vendor.
```

Comment:
I admit I did not have a chance to read the whole post (yet). Do we comment it further, e.g.:
- we propose what kind of more tests could be executed?
- we can point eager users (maybe we would need to add some documentation on tests execution) to developed tools, so they could evaluate the results for their particular system? 
- others?

Reply:
Added in https://github.com/3mdeb/news-and-ideas/pull/626/commits/6366735ed2747b36cbd5d762667d646d8519c1d2. As for tool documentation, README has a short description of the steps, I just have to add new features added recently to it.